File: Week_02/G20200343030632/Leetcode_77_632.java
import java.util.LinkedList;
import java.util.List;

class Solution {
    public List<List<Integer>> combine(int n, int k) {
        List<List<Integer>> res = new LinkedList<>();
        helper(1, n, k, new LinkedList<Integer>(), res);
        return res;
    }

    public void helper(int start, int n, int k, LinkedList<Integer> curr, List<List<Integer>> res) {
        if (curr.size() == k) {
            System.out.println(curr.toString());
            res.add(new LinkedList<>(curr));
            return ;
        }

        for (int i = start; i <= n; i++) {
            curr.add(i);
            helper(i + 1, n, k, curr, res);
            curr.removeLast();
        }

    }

    
}

File: Week_02/G20200343030632/Leetcode_77_632.py
#
# @lc app=leetcode.cn id=77 lang=python3
#
# [77] 组合
#

# @lc code=start
class Solution:
    def combine(self, n: int, k: int) -> List[List[int]]:
        if k == 0:
            return [[]]
        nums = [i for i in range(1, n + 1)]
        # if k == n:
            # return [[i for i in range(1, n + 1)]]
        res = []
        self.helper(nums, k, [], res)
        return res

    def helper(self, nums, k, current, res):
        # 退出条件以及结果处理
        if len(current) == k:
            res.append(current)
            return
        # 当前层处理
        for i, val in enumerate(nums):
            self.helper(nums[i + 1:], k, current + [val], res)
        
# @lc code=end

